### PR TITLE
Analytic continuation with arbitrary precision floating point arithmetic

### PR DESCRIPTION
**Pull request by moritzleucke**
- linked against the GMP library for arbitrary precision float support
- Thiele Pade Interpolation works now with arbitrary precision when GMP is linked
- both the greedy and non-greedy algorithms work with GMP (and also without GMP)
- added documentation
- added unit tests using zofu

### CODE REVIEW

File: GX-AnalyticContinuation/src/pade_approximant.f90 [hunk 1]
```diff
@@ -1,5 +1,8 @@
-!  Copyright (C) 2020-2022 GreenX library
+! ***************************************************************************************************
+!  Copyright (C) 2020-2023 GreenX library
```
Comment (dgolze) [inline GX-AnalyticContinuation/src/pade_approximant.f90:2]: should be 2024

File: GX-AnalyticContinuation/test/test_pade_approximant.f90 [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    real(dp), optional, intent(in) :: tol
+    ! abs() evaluates to real, hence tolerance is real
+    real(dp) :: tolerance = 1.-8_dp
+    if (present(tol)) tolerance = tol
+    is_close_complex_dp = abs(a - b) <= tolerance
+  end function is_close_complex_dp
+
+
+
+  !> Test the Pade interpolant against the function -1 / (x - x0)
```
Comment (dgolze) [inline GX-AnalyticContinuation/test/test_pade_approximant.f90:45]: doxygen headers?

File: GX-AnalyticContinuation/test/test_pade_approximant.f90 [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+    !> Clean-up
+    deallocate(x)
+    deallocate(f)
+
+  end subroutine test_pade
+
+
+
+  !> Test the GMP Pade interpolant against the function -1 / (x - x0)
```
Comment (dgolze) [inline GX-AnalyticContinuation/test/test_pade_approximant.f90:84]: doxygen header

File: GX-AnalyticContinuation/test/test_pade_approximant.f90 [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+    !> Clean-up
+    deallocate(x, f, xx, f_approx)
+    call free_params(params_thiele)
+
+  end subroutine test_pade_mp
+
+
+
+  !> Test the Thiele-Pade interpolant against the function 1 / (-x^2 + 1) which has poles
```
Comment (dgolze) [inline GX-AnalyticContinuation/test/test_pade_approximant.f90:128]: doxygen header?

File: GX-AnalyticContinuation/test/test_pade_approximant.f90 [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+    !> Clean-up
+    deallocate(x)
+    deallocate(f)
+
+  end subroutine test_thiele_pade_poles
+
+
+
+  !> Test the GMP Thiele-Pade interpolant against the function 1 / (-x^2 + 1) which has poles
```
Comment (dgolze) [inline GX-AnalyticContinuation/test/test_pade_approximant.f90:166]: header?

File: GX-AnalyticContinuation/test/test_pade_approximant.f90 [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+
+    !> Clean-up
+    deallocate(x)
+    deallocate(f)
+
+  end subroutine test_thiele_pade_abs
+
+
+
+  !> Test the GMP Thiele-Pade interpolant against the function |x| which has a branch point
```
Comment (dgolze) [inline GX-AnalyticContinuation/test/test_pade_approximant.f90:252]: see above

File: GX-AnalyticContinuation/test/test_pade_approximant.f90 [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+
+  !> Test the Pade interpolant against the function -1 / (x - x0)
+  subroutine test_pade(test)
+    !> Test object
+    class(unit_test_type), intent(inout) :: test
+
+    !> N sampling points
+    integer, parameter :: n = 100
+    !> Variable and function, respectively
+    complex(dp), allocatable :: x(:), f(:)
```
Comment (dgolze) [inline GX-AnalyticContinuation/test/test_pade_approximant.f90:53]: the style in teh rest of the library is actually complex(kind=dp) etc.

**Review by dgolze — CHANGES_REQUESTED** [review]
can you also double-check that we have the right copy-right headers (year etc.)

**Review by dgolze — APPROVED** [review]

**Review by fdelesma — APPROVED** [review]
Looks good to me. Thanks

**Pull request merged**